### PR TITLE
Decouple TritonInferenceStage from pipeline mode

### LINKED ISSUE #1378: [FEA]: Make logits a constructor argument to TritonInferenceStage
### Is this a new feature, an improvement, or a change to existing functionality?

Change

### How would you describe the priority of this feature request

Low (would be nice)

### Please provide a clear description of problem this feature solves

Currently this functionality is hard-coded to the NLP pipeline.

### Describe your ideal solution

For backwards compatibility, we can make this argument default to `True` for `TritonInferenceNLP`, and `False` for the other subclasses.

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by dagardner-nv**
## Description
* Add `needs_logits` & `inout_mapping` a constructor arguments to `TritonInferenceStage`.
* Consolidate triton inference worker impls into a single `TritonInferenceWorker` class
* For compatibility the `needs_logits` & `inout_mapping`  arguments default to default values inferred by the pipeline mode.
* Refactor custom triton inference stage from `examples/log_parsing` example as a subclass of `TritonInferenceStage` removing several lines of redundant code.
* `examples/log_parsing` example now works in C++ mode
* Add the ability to subclass a stage registered with the CLI, and register the sublcass with a new CLI name.
* Add the ability to parse dictionary types with click.
* Remove `inf-triton` from the AE mode in the CLI
* Misc linting/formatting pylint suggestions

Closes #1378

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

**Review by mdemoret-nv — APPROVED** [review]

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Make the following changes:

- Remove `AE` support from the triton stage all together
- Remove all implementations of the worker class
- Make all options of the worker class instance values (need_logits, and inout_mapping)
- Set defaults for inout_mapping similar to default for `need_logits` (i.e. by mode if not supplied)

File: morpheus/stages/inference/triton_inference_stage.py [hunk 1]
```diff
@@ -628,263 +638,22 @@ def process(self, batch: MultiInferenceMessage, cb: typing.Callable[[TensorMemor
...
-        """
+INFERENCE_WORKER_DEFAULT_INOUT_MAPPING = {
+    PipelineModes.FIL: {
+        "output__0": "probs",
+    },
+    PipelineModes.NLP: {
+        "attention_mask": "input_mask",
+        "output": "probs",
+    }
+}
```
Comment (mdemoret-nv) [inline morpheus/stages/inference/triton_inference_stage.py:653]: Should we make this a class property?

**Review by mdemoret-nv — APPROVED** [review]
A few suggestions and questions, but overall its looking good.

File: morpheus/stages/inference/triton_inference_stage.py [hunk 2]
```diff
@@ -405,9 +405,9 @@ def build_input(self, name: str, data: cp.ndarray, force_convert_inputs: bool) -
 
 
 # This class is exclusively run in the worker thread. Separating the classes helps keeps the threads separate
-class _TritonInferenceWorker(InferenceWorker):
+class TritonInferenceWorker(InferenceWorker):
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply morpheus/stages/inference/triton_inference_stage.py:408]: My thoughts were that the previous workers such as `TritonInferenceNLP` were public, then the base class they were folded into should also be public.

If the log parsing example needs to subclass `TritonInferenceWorker`, then yes, it should be public. 

If on the other-hand it can work without subclassing, then I agree it should remain private.

File: examples/log_parsing/inference.py [hunk 3]
```diff
@@ -104,13 +67,12 @@ def build_output_message(self, x: MultiInferenceMessage) -> MultiPostprocLogPars
...
-                                                        memory=memory,
-                                                        offset=0,
-                                                        count=x.count)
-        return output_message
+        return MultiPostprocLogParsingMessage(meta=x.meta,
+                                              mess_offset=x.mess_offset,
+                                              mess_count=x.mess_count,
+                                              memory=memory,
+                                              offset=0,
+                                              count=x.count)
```
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply examples/log_parsing/inference.py]: Good catch

File: morpheus/stages/inference/triton_inference_stage.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (dagardner-nv) [reply morpheus/stages/inference/triton_inference_stage.py:408]: I think it should be public, the primary reason is if a user wants a custom Triton inference worker, currently the only public class is the `InferenceWorker` requiring the user to replicate code currently in the `_TritonInferenceWorker` class.

With that said, the counter-argument is that I think the primary reason a user would need a custom worker would be to handle tensors in addition to the `probs` tensor. Which could be a constructor argument, which I believe would allow us to remove the custom stage from the log parsing example.

**Review by mdemoret-nv — APPROVED** [review]
Lets get these changes merged but make an issue before you do to try and remove the inference worker class entirely.
